test/unit/core/bit_manipulation_test.cpp

```diff
@@ -108,6 +108,103 @@ TYPED_TEST(unsigned_operations, bit_scan_reverse)
     }
 }
 
+TYPED_TEST(unsigned_operations, count_leading_zeros)
+{
+    using unsigned_t = TypeParam;
+    constexpr size_t t1 = count_leading_zeros<unsigned_t>(0b0001);
+    constexpr size_t t2 = count_leading_zeros<unsigned_t>(0b0101);
+    constexpr size_t t3 = count_leading_zeros<unsigned_t>(0b0010);
+    constexpr size_t t4 = count_leading_zeros<unsigned_t>(0b0110);
+    constexpr size_t t5 = count_leading_zeros<unsigned_t>(0b0100);
+    constexpr size_t t6 = count_leading_zeros<unsigned_t>(0b10100000);
+    EXPECT_EQ(t1, sizeof_bits<unsigned_t> - 1u);
+    EXPECT_EQ(t2, sizeof_bits<unsigned_t> - 3u);
+    EXPECT_EQ(t3, sizeof_bits<unsigned_t> - 2u);
+    EXPECT_EQ(t4, sizeof_bits<unsigned_t> - 3u);
+    EXPECT_EQ(t5, sizeof_bits<unsigned_t> - 3u);
+    EXPECT_EQ(t6, sizeof_bits<unsigned_t> - 8u);
+
+    for (uint8_t cnt = 0; cnt < sizeof_bits<unsigned_t>; ++cnt)
+    {
+        unsigned_t start = std::numeric_limits<unsigned_t>::max() >> cnt;
+        unsigned_t end = start >> 1u;
+        for (unsigned_t n = start, k = 0u; n < end && k < max_iterations; ++n, ++k)
+        {
+            EXPECT_EQ(count_leading_zeros(n), cnt) << "n " << n << " should have " << cnt << " leading zeros.";
+        }
+    }
+}
+
+TYPED_TEST(unsigned_operations, count_trailing_zeros)
+{
+    using unsigned_t = TypeParam;
+    constexpr size_t zero  = count_trailing_zeros<unsigned_t>(0b0001);
+    constexpr size_t zero2 = count_trailing_zeros<unsigned_t>(0b0101);
+    constexpr size_t one1  = count_trailing_zeros<unsigned_t>(0b0010);
+    constexpr size_t one2  = count_trailing_zeros<unsigned_t>(0b0110);
+    constexpr size_t two   = count_trailing_zeros<unsigned_t>(0b0100);
+    constexpr size_t five  = count_trailing_zeros<unsigned_t>(0b10100000);
+    EXPECT_EQ(zero,  0u);
+    EXPECT_EQ(zero2, 0u);
+    EXPECT_EQ(one1,  1u);
+    EXPECT_EQ(one2,  1u);
+    EXPECT_EQ(two,   2u);
+    EXPECT_EQ(five,  5u);
+
+    for (uint8_t cnt = 0; cnt < sizeof_bits<unsigned_t>; ++cnt)
+    {
+        unsigned_t start = std::numeric_limits<unsigned_t>::max() << cnt;
+        unsigned_t end = start << 1u;
+        for (unsigned_t n = start, k = 0u; n < end && k < max_iterations; ++n, ++k)
+        {
+            EXPECT_EQ(count_trailing_zeros(n), cnt) << "n " << n << " should have " << cnt << " trailing zeros.";
+        }
+    }
+}
+
+// http://graphics.stanford.edu/~seander/bithacks.html#NextBitPermutation
+template <std::UnsignedIntegral unsigned_t>
+unsigned_t permute_bits(unsigned_t v)
+{
+    if (v & (unsigned_t{1u} << (sizeof_bits<unsigned_t> - 1)))
```

Comment:
As a note:
The function produces permutations with the same number of 1 bits, e.g.
```
0b0000'0001
0b0000'0010
0b0000'0100
...
```
For some reason, for `uint32_t` and `uint64_t`, permuting `0b1000'00..` results in setting all bits to 1, even though it probably should not (http://coliru.stacked-crooked.com/a/e445e8159c8e45ce). had this problem locally and on CI. So I added the if to catch this.

Reply:
The compiler might changes it to:

http://coliru.stacked-crooked.com/a/9d3b79279787b6ce

Which produces 111111111111111111.

Reply:
The method of http://graphics.stanford.edu/~seander/bithacks.html#NextBitPermutation does not seem to work when the msb is set. So the fix seems fine.